### ISSUE TITLE
scheduler interface

### ISSUE THREAD

**Issue by ekinakyurek** [denizyuret/Knet.jl#564]
For reference PyTorch interface:
```Python
optimizer = optim.SGD(net.parameters(), lr=opt.learning_rate, momentum=0.9, weight_decay=1e-4)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=10, gamma=0.2) 

for epoch in range(2):  # loop over the dataset multiple times
    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):
        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data
        optimizer.zero_grad()
        outputs = net(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
    scheduler.step() # after each epoch

```

**Comment by andevellicus** [comment]
Not sure if this would be helpful, but I implemented this in my code:

``` 
# callbacks.jl
using DelimitedFiles

# Struct to hold the training state
struct TrainingState
    current_epoch::Int
    model::Model
    train_loss::AbstractFloat
    test_metrics::Vector{AbstractFloat}
end

#=  
A functor that implements the AbstractCallback type should
should take a TrainingState as an argument:

struct Callback <: AbstractCallback
    ...fields...
end

function(cb::Callback)(state::TrainingState)
    ...do things...
end
=#
abstract type AbstractCallback end

#= 
CSVLogger callback constructor. Will log metrics to a file

Arguments
* logdir: directory of the logfiles
* filename: logfile name
* metrics: Array of symbols representing which metrics to track 
=#
struct CSVLogger <: AbstractCallback
    filepath::String
end

function CSVLogger(logdir::String, filename::String, metrics::Array{Symbol})
    if !isdir(logdir)
        mkdir(logdir)
    end
    headers = ["test_" * string(i) for i in metrics]
    pushfirst!(headers, "train_loss")
    pushfirst!(headers, "epoch")
    headers = reshape(headers, 1, :)
    filepath = logdir * "/" * filename
    open(filepath, "w") do io
        writedlm(io, headers, ',')
    end
    CSVLogger(filepath)
end

function(logger::CSVLogger)(state::TrainingState)
    row = vcat(state.current_epoch, state.train_loss, state.test_metrics)'
    open(logger.filepath, "a") do io
        writedlm(io, row, ',')
    end
end

#= 
Model checkpoint callback. Will save models

Arguments
* save_path: Model save path
* save_best: If true, will only save the best model
* mode: If "min", we expect the best_test_score to decrease. If
        "max," we expect the best test score to increase. Ignored
        if save_best is false
* best_test_score: Keeps track of what our best test score is
  ** NB: For this and for LRDecay, the tracked test score is assumed to be
        TrainingState.test_metrics[1] 
=#
mutable struct ModelCheckpoint <: AbstractCallback
    save_path::String
    save_best::Bool
    mode::String
    best_test_score::AbstractFloat
end

function ModelCheckpoint(path::String, mode::String; save_best=true)
    if mode == "max"
        ModelCheckpoint(path, save_best, mode, 0.0)
    elseif mode == "min"
        ModelCheckpoint(path, save_best, mode, 1.0)
    elseif save_best
        throw(ArgumentError("'Mode' argument must be either 'min' or 'max' if save_best is true"))
    end
end

function(cp::ModelCheckpoint)(state::TrainingState)
    function save_best_model()
        cp.best_test_score = state.test_metrics[1]
        @info "Saving best model"
        Knet.save(cp.save_path, "model", state.model)
    end

    # Again, the tracked metric is assumted to be test_metrics[1]
    if cp.mode == "max" && cp.save_best && state.test_metrics[1] > cp.best_test_score
        save_best_model()
    elseif cp.mode == "min" && cp.save_best && state.test_metrics[1] < cp.best_test_score
        save_best_model()
    elseif !cp.save_best
        @info "Saving model"
        Knet.save(cp.save_path, "model", state.model)
    end
end

#=
Learning Rate decay callback. Reduces the LR if the training metric hasn't improved.

Arguments:
* min_lr: Minimum learning rate
* factor: by how much to reduce the learning rate
* patience: how many epochs of no improvement should we wait
* mode: do we expect the training metric to decrease (min) or increase (max)
* best_test_score: Keeps track of what our best test score is
  ** NB: For this and for ModelCheckpoint, the tracked test score is assumed to be
        TrainingState.test_metrics[1] 
* no_improve_count: counter that tracks how many epochs without improvement    
=#
mutable struct LRDecay <: AbstractCallback
    min_lr::Float32
    factor::Float32
    patience::Int
    mode::String
    best_test_score::AbstractFloat
    plateau_counter::Int
end

function LRDecay(min_lr::Float32, factor::Float32, patience::Int, mode::String)
    if mode == "max"
        LRDecay(min_lr, factor, patience, mode, 0.0, 0)
    elseif mode == "min"
        LRDecay(min_lr, factor, patience, mode, 1.0, 0)
    else
        throw(ArgumentError("'Mode' argument must be either 'min' or 'max'"))
    end
end

function(d::LRDecay)(state::TrainingState)
    function decay()
        # Iterate the counter
        d.plateau_counter += 1
        # If we haven't improved...
        if d.plateau_counter > d.patience
            new_lr = -1
            # Update the LR
            for p in params(state.model)
                if p.opt.lr*d.factor <= d.min_lr
                    new_lr = p.opt.lr = d.min_lr
                else
                    new_lr = p.opt.lr = p.opt.lr*d.factor
                end
            end
            @info "LR set to $new_lr"
            # Reset the counter
            d.plateau_counter = 0
        end
    end

    # Again, the tracked metric is assumted to be test_metrics[1]
    if d.mode == "max"
        if state.test_metrics[1] > d.best_test_score
            d.best_test_score = state.test_metrics[1]
            d.plateau_counter = 0
        else
            decay()
        end
    elseif d.mode == "min"
        if state.test_metrics[1] < d.best_test_score
            d.best_test_score = state.test_metrics[1]
            d.plateau_counter = 0
        else
            decay()
        end
    end
end

#=
Learning Rate cycler callback. Slowly ramps up the LR to a midway point and then
decrease it back down, then annihilates the LR.

Arguments:
* target_lr: Target learning rate
* init_lr: Initial LR.
* n_epochs: Number of epochs. Used to calculate a midpoint 
* split: Used to split the number of epochs into a ramping up and ramping down portion.
=#
mutable struct LRCycler <: AbstractCallback
    midpoint::Int
    multiplier::Float32
    current_lr::Float32
end

function LRCycler(target_lr::Float32, init_lr::Float32, n_epochs::Int; split=0.85f0)
    midpoint = Int(round(n_epochs * split / 2))
    multiplier = (target_lr/init_lr) ^ (1/midpoint)
    LRCycler(midpoint, Float32(multiplier), init_lr)
end

function (c::LRCycler)(state::TrainingState)
    if state.current_epoch < c.midpoint
   	new_lr = c.current_lr * c.multiplier
    elseif state.current_epoch < c.midpoint * 2
   	new_lr = c.current_lr / c.multiplier
    else
	new_lr = c.current_lr * 0.1
    end
    
    for p in params(state.model); p.opt.lr = new_lr; end
    
    @info "LR set to $new_lr"
    c.current_lr = new_lr
end
```

Then in train.jl:
```
# init stuff...

    # Define our callbacks
    callbacks = [CSVLogger("logs", string(DateTime(now())) * "_" * 
			   string(typeof(model)) * "_" * 
			   o[:model_name] * ".log", [:dice]),
		 ModelCheckpoint(string(typeof(model)) * "_" *

# ... define more stuff....
# ... train loop....

# At the end of train loop do:
	state = TrainingState(i, model, mean(train_losses), [avg_test_dice])

	# Do callbacks
	for cb in callbacks
	    if isa(cb, AbstractCallback); cb(state); end
	end
```
